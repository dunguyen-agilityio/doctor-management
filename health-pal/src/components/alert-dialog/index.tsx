import { AlertDialog, AlertDialogContentProps, AlertDialogProps, TamaguiElement } from 'tamagui'

import { WINDOW_SIZE } from '@app/constants'

import { useModal } from '@app/hooks/use-modal'

import { Button, Heading, Text, XStack } from '@app/components/common'

import { ModalRef } from '@app/types/modal'

interface CustomAlertDialogProps extends AlertDialogProps {
  title?: string
  description?: string
  cancelable?: boolean
  onConfirm?: () => void
  actionTitle?: string
  ref?: React.Ref<ModalRef | null>
  trigger?: React.ReactNode
  container?: ({ children }: React.PropsWithChildren) => React.ReactNode
  contentProps?: AlertDialogContentProps & React.RefAttributes<TamaguiElement>
}

const CustomAlertDialog = ({
  title,
  description,
  onConfirm,
  cancelable,
  actionTitle = 'OK',
  children,
  ref,
  container: Container = ({ children }) => <>{children}</>,
  trigger = <AlertDialog.Trigger asChild />,
  contentProps,
  ...props
}: React.PropsWithChildren<CustomAlertDialogProps>) => {
  const [open, setOpen] = useModal(ref)

  return (
    <AlertDialog open={open} onOpenChange={setOpen} {...props}>
      {trigger}
      <AlertDialog.Portal>
        <AlertDialog.Overlay
          key="overlay"
          animation="quick"
          opacity={0.5}
          enterStyle={{ opacity: 0 }}
          exitStyle={{ opacity: 0 }}
          aria-hidden
        />

        <AlertDialog.Content
          role="alertdialog"
          elevate
          key="content"
          animation={[
            'quick',
            {
              opacity: {
                overshootClamping: true,
              },
            },
          ]}
          enterStyle={{ x: 0, y: -20, opacity: 0, scale: 0.9 }}
          exitStyle={{ x: 0, y: 10, opacity: 0, scale: 0.95 }}
          x={0}
          scale={1}
          opacity={1}
          width={WINDOW_SIZE.width - 48}
          borderRadius={8}
          paddingHorizontal={20}
          paddingVertical={12}
          aria-labelledby="alert-dialog-title"
          accessibilityHint={description ? 'alert-dialog-description' : undefined}
          gap={12}
          {...contentProps}>
          {title && (
            <AlertDialog.Title>
              <Heading size="large">{title}</Heading>
            </AlertDialog.Title>
          )}
          {description && (
            <AlertDialog.Description>
              <Text size="small">{description}</Text>
            </AlertDialog.Description>
          )}
          {children}
          <XStack gap={12} justifyContent="flex-end">
            {cancelable && (
              <AlertDialog.Cancel asChild>
                <Button variant="secondary" aria-label="Cancel dialog" testID="cancel-button">
                  Cancel
                </Button>
              </AlertDialog.Cancel>
            )}
            {onConfirm && (
              <AlertDialog.Action asChild>
                <Button
                  onPress={onConfirm}
                  aria-label={`${actionTitle} action`}
                  testID="confirm-button">
                  {actionTitle}
                </Button>
              </AlertDialog.Action>
            )}
          </XStack>
        </AlertDialog.Content>
      </AlertDialog.Portal>
    </AlertDialog>
  )
}

export const AlertDialogDescription = AlertDialog.Description

export const AlertDialogTrigger = AlertDialog.Trigger

export const AlertDialogTitle = AlertDialog.Title

export const AlertDialogAction = AlertDialog.Action

export const AlertDialogCancel = AlertDialog.Cancel
export const AlertDialogOverlay = AlertDialog.Overlay
export const AlertDialogContent = AlertDialog.Content
export const AlertDialogPortal = AlertDialog.Portal
export const AlertDialogClose = AlertDialog.Cancel

export default CustomAlertDialog
