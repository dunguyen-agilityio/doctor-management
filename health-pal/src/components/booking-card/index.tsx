import { useRef } from 'react'

import { router } from 'expo-router'

import { Card, Separator } from 'tamagui'

import { PLACEHOLDER_IMAGE } from '@app/constants/image'

import { Button, Heading, Text, XStack, YStack } from '@theme'

import { BOOKING_TABS } from '@app/types/booking'
import { ModalRef } from '@app/types/modal'
import CancelBookingModal from '@app/ui/booking/cancel-booking-modal'

import CloudinaryImage from '../cloudinary-image'

type TBookingCard = {
  date: string
  time: string
  doctorName: string
  doctorAvatar?: string
  specialty: string
  address: string
  documentId: string
  doctorId: string
}

interface BookingCardProps extends TBookingCard {
  type?: BOOKING_TABS
}

const BookingCard = ({
  date,
  time,
  doctorName,
  specialty,
  doctorAvatar,
  address,
  documentId,
  doctorId,
  ...props
}: BookingCardProps) => {
  const cancelBookRef = useRef<ModalRef>(null)
  const { type = BOOKING_TABS.UPCOMING } = props

  const renderAction = () => {
    const ACTIONS: Record<BOOKING_TABS, { title: string; action: () => void }[]> = {
      [BOOKING_TABS.CANCELED]: [],
      [BOOKING_TABS.COMPLETED]: [
        {
          title: 'Re-Book',
          action: () => router.navigate({ pathname: '/(app)/booking', params: { doctorId } }),
        },
        {
          title: 'Add Review',
          action: () =>
            router.navigate({ pathname: '/(app)/doctors/details/[id]', params: { id: doctorId } }),
        },
      ],
      [BOOKING_TABS.UPCOMING]: [
        {
          title: 'Cancel',
          action: () => {
            cancelBookRef.current?.open()
          },
        },
        {
          title: 'Reschedule',
          action: () =>
            router.navigate({
              pathname: '/(app)/booking',
              params: { doctorId, bookingId: documentId, date, time },
            }),
        },
      ],
    }

    const actions = ACTIONS[type]

    if (actions.length < 1) return null

    const [action1, action2] = actions

    return (
      <XStack marginTop={10} gap={10}>
        <Button variant="secondary" sizeButton="sm" flex={1} onPress={action1.action}>
          {action1.title}
        </Button>
        <Button flex={1} sizeButton="sm" onPress={action2.action}>
          {action2.title}
        </Button>
      </XStack>
    )
  }
  return (
    <>
      <Card
        borderRadius={12}
        borderWidth={0.5}
        borderColor="$grey100"
        backgroundColor="$white"
        shadowColor="$black"
        shadowOffset={{ width: 0, height: 4 }}
        shadowOpacity={0.1}
        shadowRadius={6}
        elevation={3}
        padding={10}
        marginBottom={10}>
        {/* Date and Time */}
        <Card.Header>
          <Heading fontSize={14}>
            {date} - {time}
          </Heading>
          <Separator marginVertical={12} />
        </Card.Header>

        <Card.Footer>
          {/* Doctor Info */}
          <YStack flex={1}>
            <XStack gap={10}>
              <CloudinaryImage
                source={{ uri: doctorAvatar }}
                style={{ height: 109, width: 109, borderRadius: 8 }}
                placeholder={PLACEHOLDER_IMAGE}
              />
              <YStack paddingVertical={14}>
                <Text fontSize={16} fontWeight="bold">
                  {doctorName}
                </Text>
                <Text fontSize={14} color="#6B7280">
                  {specialty}
                </Text>
                <Text fontSize={12} color="#9CA3AF" numberOfLines={1} maxWidth={250}>
                  {address}
                </Text>
              </YStack>
            </XStack>

            {/* Buttons */}
            {renderAction()}
          </YStack>
        </Card.Footer>
      </Card>
      {props.type === BOOKING_TABS.UPCOMING && (
        <CancelBookingModal
          ref={cancelBookRef}
          date={date}
          doctorName={doctorName}
          documentId={documentId}
          time={time}
        />
      )}
    </>
  )
}

export default BookingCard
