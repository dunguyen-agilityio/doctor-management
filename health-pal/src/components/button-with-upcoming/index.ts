import { Button } from '@theme'

import { withUpcomingFeature } from '@app/hocs/with-upcoming-feature'

const ButtonWithUpcoming = withUpcomingFeature(Button)

export default ButtonWithUpcoming
