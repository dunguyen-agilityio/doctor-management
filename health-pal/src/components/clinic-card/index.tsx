import { Image } from 'expo-image'

import { Card, Separator, Stack } from 'tamagui'

import { Heading } from '@theme/heading'
import { XStack, YStack } from '@theme/stack'
import { Text } from '@theme/text'

import { Hospital, Routing } from '@icons'
import LocationOutline from '@icons/location-outline'

import Stars from '@app/components/stars'
import useMediaQuery, { MediaQuery } from '@app/hooks/use-media-query'
import { Clinic } from '@app/models/clinic'
import { FAVORITE_TYPES } from '@app/types/favorite'
import FavoriteButton from '@app/ui/favorite/favorite-button'

import { tokens } from '@/tamagui.config'

interface ClinicCardProps extends Clinic, MediaQuery {
  favoriteId?: string
}

const ClinicCard = ({
  image = require('@/assets/images/banner01.webp'),
  address,
  name,
  rating = 0,
  reivewCouter = 0,
  w = 342,
  type,
  px,
  full,
  h,
  favoriteId,
  id,
  documentId,
}: ClinicCardProps) => {
  const { width, height } = useMediaQuery({ w, px, h, full })

  return (
    <Stack position="relative">
      <FavoriteButton
        color={tokens.color.white.val}
        type={FAVORITE_TYPES.HOSPITAL}
        clinicId={id}
        clinicName={name}
        favoriteId={favoriteId}
        position="absolute"
        zIndex={100}
        top={6}
        right={6}
      />
      <Card
        elevate
        bordered
        width={width}
        height={height}
        borderRadius="$2"
        overflow="hidden"
        shadowColor="$white">
        <Card.Header padding={0}>
          <Image source={image?.url} style={{ height: 120, objectFit: 'contain' }} />
        </Card.Header>

        <Card.Footer padding={12} paddingTop={8}>
          <YStack gap="$sm">
            <Heading size="small" numberOfLines={1}>
              {name}
            </Heading>

            <XStack alignItems="center" gap="$sm">
              <LocationOutline />
              <Text size="extraSmall" numberOfLines={1}>
                {address}
              </Text>
            </XStack>

            <XStack alignItems="center" gap="$sm">
              <Stars stars={rating} max={1} />
              <Text size="extraSmall">{`(${reivewCouter} Reviews)`}</Text>
            </XStack>

            <Separator backgroundColor="$grey200" />

            <XStack justifyContent="space-between">
              <XStack alignItems="center" gap="$sm">
                <Routing />
                <Text size="extraSmall">2.5 km/40min</Text>
              </XStack>

              <XStack alignItems="center" gap="$sm">
                <Hospital />
                <Text size="extraSmall">{type}</Text>
              </XStack>
            </XStack>
          </YStack>
        </Card.Footer>
      </Card>
    </Stack>
  )
}

export default ClinicCard
