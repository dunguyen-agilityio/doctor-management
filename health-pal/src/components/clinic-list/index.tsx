import { FlatList } from 'react-native-gesture-handler'

import { Stack } from 'tamagui'

import { YStack } from '@theme'

import ClinicCard from '@app/components/clinic-card'
import SessionHeader from '@app/components/session-header'
import { Clinic } from '@app/models/clinic'

const ItemSeparatorComponent = () => <Stack width={16} />

const keyExtractor = (item: Clinic) => item.documentId

const ClinicList = ({ data }: { data: Clinic[] }) => {
  const renderItem = ({ item }: { item: Clinic }) => <ClinicCard w={232} px={0} {...item} />

  return (
    <YStack gap={10} paddingBottom={16}>
      <SessionHeader title="Nearby Medical Centers" />
      <FlatList
        data={data}
        renderItem={renderItem}
        ItemSeparatorComponent={ItemSeparatorComponent}
        keyExtractor={keyExtractor}
        horizontal
        showsHorizontalScrollIndicator={false}
        showsVerticalScrollIndicator={false}
      />
    </YStack>
  )
}

export default ClinicList
