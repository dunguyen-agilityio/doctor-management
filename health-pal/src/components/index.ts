export * from './alert-dialog'
export { default as AlertDialog } from './alert-dialog'
export { default as ButtonWithUpcoming } from './button-with-upcoming'
export { default as DateInput } from './date-input'
export { default as DatePicker } from './date-picker'
export { default as DoctorCard } from './doctor-card'
export { default as FormKeyboardAvoidingView } from './form-keyboard-avoiding-view'
export { default as HospitalCard } from './hospital-card'
export { default as Input } from './input'
export { default as LoadingIndicator } from './loading-indicator'
export { default as Modal } from './modal'
export { default as ReviewCard } from './review-card'
export { default as SearchInput } from './search-input'
export { default as Select } from './select'
export { default as SessionHeader } from './session-header'
export { default as Stat } from './stat'
export * from './tab'
export { default as Upload } from './upload'
