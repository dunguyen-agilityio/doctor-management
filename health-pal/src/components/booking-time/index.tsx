import { StyleSheet } from 'react-native'

import { TIME_SLOTS } from '@/constants/booking'

import { Heading, Text, XStack, YStack } from '@/components/common'

import TimeButton from '../time-button'

interface BookingTimeProps {
  errorMessage?: string
  onChange: (value: string) => void
  value?: string
  disable?: (time: string) => boolean
}

const BookingTime = ({ errorMessage, onChange, value, disable }: BookingTimeProps) => {
  return (
    <YStack gap={8}>
      <XStack justifyContent="space-between" alignItems="center">
        <Heading>Select Hour</Heading>
        {errorMessage && (
          <Text size="extraSmall" color="red">
            {errorMessage}
          </Text>
        )}
      </XStack>
      <XStack justifyContent="space-between" gap={14} flexWrap="wrap">
        {TIME_SLOTS.map((time) => {
          const disabled = disable?.(time)

          const backgroundColor = !disabled && value === time ? '$primary' : '$grey50'
          const color = !disabled && value === time ? '$white' : '$primary'

          return (
            <TimeButton
              value={time}
              key={time}
              onSelect={onChange}
              disabled={disabled}
              color={color}
              backgroundColor={backgroundColor}
              disabledStyle={styles.disabledStyle}
            />
          )
        })}
      </XStack>
    </YStack>
  )
}

export default BookingTime

const styles = StyleSheet.create({
  disabledStyle: { opacity: 0.5, backgroundColor: '$grey300' },
})
