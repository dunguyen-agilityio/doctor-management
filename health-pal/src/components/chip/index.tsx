import { GestureResponderEvent } from 'react-native'

import { Button, StackProps, TextProps } from 'tamagui'

import { Text } from '@/components/common'

type Size = 'md'

interface ChipProps extends StackProps {
  active?: boolean
  children: React.ReactNode
  size?: Size
  onSelect: (value: string) => void
  value: string
}

const STYLES_BY_SIZE: Record<
  Size,
  {
    container: (theme: { active: boolean }) => StackProps
    text: (theme: { active: boolean }) => TextProps
  }
> = {
  md: {
    container: ({ active }) => ({
      backgroundColor: active ? '$primary' : '$white',
      borderRadius: 60,
      borderColor: '$primary',
      borderWidth: 1,
      height: 37,
      paddingHorizontal: 20,
      paddingVertical: 8,
    }),
    text: ({ active }) => ({
      color: active ? '$white' : '$primary',
    }),
  },
}

const Chip = ({
  active = false,
  children,
  value,
  size = 'md',
  onSelect,
  onPress,
  ...props
}: ChipProps) => {
  const { container, text } = STYLES_BY_SIZE[size]

  let content = children

  if (typeof children === 'string') {
    content = (
      <Text size="small" fontWeight="600" {...text({ active })}>
        {children}
      </Text>
    )
  }

  const handlePress = (event: GestureResponderEvent) => {
    onPress?.(event)
    onSelect(value)
  }

  return (
    <Button {...container({ active })} {...props} onPress={handlePress}>
      {content}
    </Button>
  )
}

export default Chip
