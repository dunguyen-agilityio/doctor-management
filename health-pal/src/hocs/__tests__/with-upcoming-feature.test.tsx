import { act, fireEvent, render, screen } from '@utils-test'

import { ButtonProps } from 'tamagui'

import { Button } from '@/components'

import { withUpcomingFeature } from '../with-upcoming-feature'

const TestComponent = ({ onPress, children }: ButtonProps) => (
  <Button testID="test-button" onPress={onPress}>
    {children}
  </Button>
)

const TestComponentWithUpcoming = withUpcomingFeature(TestComponent)

describe('withUpcomingFeature', () => {
  beforeEach(() => {
    jest.clearAllMocks()
  })

  afterEach(() => {
    jest.restoreAllMocks()
  })

  const setup = () => render(<TestComponentWithUpcoming>Test Button</TestComponentWithUpcoming>)

  it('renders wrapped component with passed props', () => {
    setup()
    const button = screen.getByTestId('test-button')
    expect(button).toBeTruthy()
    expect(button).toHaveTextContent('Test Button')
  })

  it('renders AlertDialog when onPress is triggered', () => {
    setup()
    const button = screen.getByTestId('test-button')
    act(() => {
      fireEvent.press(button)
    })
    expect(screen.getByTestId('alert-dialog-upcoming')).toBeTruthy()
  })

  it('matches snapshot', () => {
    const { toJSON } = setup()
    expect(toJSON()).toMatchSnapshot()
  })
})
