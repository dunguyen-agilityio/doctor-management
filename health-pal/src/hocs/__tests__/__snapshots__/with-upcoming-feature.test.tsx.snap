// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`withUpcomingFeature matches snapshot 1`] = `
[
  <View
    aria-label="Notifications (F8)"
    pointerEvents="box-none"
    role="region"
    style={
      {
        "bottom": 0,
        "flexDirection": "column",
        "left": 0,
        "maxWidth": "100%",
        "position": "absolute",
        "right": 0,
        "top": 0,
        "zIndex": 100000,
      }
    }
    tabIndex={-1}
  >
    <View
      focusable={false}
      pointerEvents="box-none"
      style={
        {
          "flexDirection": "column",
          "maxWidth": "100%",
          "position": "absolute",
        }
      }
    />
  </View>,
  <View
    focusVisibleStyle={{}}
    focusable={true}
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    role="button"
    style={
      {
        "alignItems": "center",
        "backgroundColor": "#1c2a3a",
        "borderBottomColor": "transparent",
        "borderBottomLeftRadius": 60,
        "borderBottomRightRadius": 60,
        "borderBottomWidth": 1,
        "borderLeftColor": "transparent",
        "borderLeftWidth": 1,
        "borderRightColor": "transparent",
        "borderRightWidth": 1,
        "borderStyle": "solid",
        "borderTopColor": "transparent",
        "borderTopLeftRadius": 60,
        "borderTopRightRadius": 60,
        "borderTopWidth": 1,
        "color": "#ffffff",
        "flexDirection": "row",
        "flexWrap": "nowrap",
        "fontWeight": "500",
        "height": 48,
        "justifyContent": "center",
        "opacity": 1,
        "paddingBottom": 12,
        "paddingLeft": 20,
        "paddingRight": 20,
        "paddingTop": 12,
      }
    }
    testID="test-button"
  >
    <Text
      lineBreakMode="clip"
      numberOfLines={1}
      style={
        {
          "color": "#ffffff",
          "flexGrow": 0,
          "flexShrink": 1,
          "fontFamily": "Inter_500Medium",
          "fontSize": 14,
          "lineHeight": 21,
          "userSelect": "none",
        }
      }
      suppressHighlighting={true}
    >
      Test Button
    </Text>
  </View>,
]
`;
